packages/langium-cli/src/generate.ts

```diff
@@ -295,21 +295,27 @@ export async function generateTypes(options: ExtractTypesOptions): Promise<void>
         }
     }
 
-    const grammarFileUri = URI.file(grammarPath);
-    const document = documents.getOrCreateDocument(grammarFileUri);
-    const allUris = eagerLoad(document);
+    const grammarDoc = await doLoadAndUpdate(documents.getOrCreateDocument(URI.file(grammarPath)));
+    const genTypes =  generateTypesFile(grammarServices, [grammarDoc.parseResult.value as Grammar]);
+    await writeWithFail(typesFilePath, genTypes, { watch: false });
+    log('log', { watch: false }, `Generated type definitions to: ${chalk.white.bold(typesFilePath)}`);
+    return;
+}
+
+/**
+ * Builds the given grammar document and all imported grammars.
+ */
+async function doLoadAndUpdate(grammarDoc: LangiumDocument): Promise<LangiumDocument> {
+    const allUris = eagerLoad(grammarDoc);
     await sharedServices.workspace.DocumentBuilder.update(allUris, []);
```

Comment:
does this call of `update` make sense? ... also wrt. the loop subsequently

Reply:
I think the `update` call does make sense.

In my case I do not have the  `LangiumConfig`. And yes, `doLoadAndUpdate ` was inspired by `buildAll` func :-)

Reply:
> In my case I do not have the `LangiumConfig`. And yes, `doLoadAndUpdate ` was inspired by `buildAll` func :-)

Right. But maybe you could put this https://github.com/langium/langium/blob/2e6acf43c3281c26e1c6788cc62f4ae7ee0884b3/packages/langium-cli/src/generate.ts#L164-L172 of  `buildAll` into a dedicated method taking a varargs parameter of URI or so.

`buildAll` might than call this method with `config.languages.map(lang => URI.file(path.resolve(relPath, languageConfig.grammar)))`.

🙏

Reply:
> What do @msujew and @spoenemann think?

Consolidating is fine if it doesn't increase the code complexity too much.

Reply:
@spoenemann 
I thought about re-using existing implementation, but decided to leave it as is, especially because I need a return value in my case. 
I'm also just about to fix a bug in the current implementation and don't want to do much changes on existing code.

@sailingKieler 
I could provide an other optimization PR which address your concerns about the code style. Doing so we can better discuss the outcome in sense of "complexity vs. duplicate code". WDYT?

Reply:
👍  reusing code can sometimes increase complexity instead of reducing it.

This is internal code, not public API, so we can always refactor it as we see fit.